packages/flutter/test/widgets/scrollbar_test.dart

```diff
@@ -0,0 +1,435 @@
+      final ScrollMetrics startingMetrics = defaultMetrics.copyWith(
+        maxScrollExtent: maxExtent,
+        viewportDimension: viewportDimension,
+      );
+      const Size size = Size(600, viewportDimension);
+      const double minLen = 0;
+
+      painter = _buildPainter(
+        minLength: minLen,
+        minOverscrollLength: minLen,
+        scrollMetrics: defaultMetrics
+      );
+
+      final List<ScrollMetrics> metricsList =
+      <ScrollMetrics> [startingMetrics.copyWith(pixels: 0.01)]
+      ..addAll(List<ScrollMetrics>.generate(
+          (maxExtent/viewportDimension).round(),
+          (int index) => startingMetrics.copyWith(pixels: (index + 1) * viewportDimension)
+        ).where((ScrollMetrics metrics) => !metrics.outOfRange))
+      ..add(startingMetrics.copyWith(pixels: maxExtent - 0.01));
+
+      double lastCoefficient;
+      for(ScrollMetrics metrics in metricsList) {
+        painter.update(metrics, metrics.axisDirection);
+        painter.paint(testCanvas, size);
+
+        final double newCoefficient = metrics.pixels/testCanvas.rect.top;
+        lastCoefficient ??= newCoefficient;
+
+        expect(testCanvas.rect.top >= 0, true);
+        expect(testCanvas.rect.bottom <= maxExtent, true);
+        expect(testCanvas.rect.left, size.width - _kThickness);
+        expect(testCanvas.rect.width, _kThickness);
+        expect(nearEqual(testCanvas.rect.height, viewportDimension * viewportDimension / (viewportDimension + maxExtent), 0.001), true);
+        expect(nearEqual(lastCoefficient, newCoefficient, 0.001), true);
+
+        testCanvas.reset();
+      }
+    }
+  );
+
+  test(
+    'mainAxisMargin is respected',
+    () {
+      const double viewportDimension = 23;
+      const double maxExtent = 100;
+      final ScrollMetrics startingMetrics = defaultMetrics.copyWith(
+        maxScrollExtent: maxExtent,
+        viewportDimension: viewportDimension,
+      );
+      const Size size = Size(600, viewportDimension);
+      const double minLen = 0;
+
+      const List<double> margins = <double> [-10, 1, viewportDimension/2 - 0.01];
+      for(double margin in margins) {
+        painter = _buildPainter(
+          mainAxisMargin: margin,
+          minLength: minLen,
+          minOverscrollLength: minLen,
+          scrollMetrics: defaultMetrics
+        );
+
+        // Overscroll to double.negativeInfinity (top).
+        painter.update(
+          startingMetrics.copyWith(pixels: double.negativeInfinity),
+          startingMetrics.axisDirection
+        );
+
+        painter.paint(testCanvas, size);
+        expect(testCanvas.rect.top, margin);
+
+        // Overscroll to double.infinity (down).
+        testCanvas.reset();
+        painter.update(
+          startingMetrics.copyWith(pixels: double.infinity),
+          startingMetrics.axisDirection
+        );
+
+        painter.paint(testCanvas, size);
+        expect(size.height - testCanvas.rect.bottom, margin);
+
+        testCanvas.reset();
```

Comment:
You wouldn't have to do this with mockito because each interaction you verify is taken off the queue. You can also use our mock_canvas